OpenRA.Mods.Common/Traits/Buildable.cs

```diff
@@ -62,6 +66,22 @@ public static string GetInitialFaction(ActorInfo ai, string defaultFaction)
 		{
 			return ai.TraitInfoOrDefault<BuildableInfo>()?.ForceFaction ?? defaultFaction;
 		}
+
+		public string GetIconSequence(Player renderPlayer, string image)
```

Comment:
Also, as this will be called every tick for every actor on production palette, it might be worth to somehow cache this. I'd assume that searching sequences and chrome metrics is not exactly cheap. I'm not sure how or if we even want to optimise though

Reply:
As for `ChromeMetrics` that should be fine as retrieval using `TryGet<T>` is cached using `Dictionary`:

```
public static bool TryGet<T>(string key, out T result)
{
	if (!data.TryGetValue(key, out var s))
	{
		result = default;
		return false;
	}

	result = FieldLoader.GetValue<T>(key, s);
	return true;
}
```

Same goes to `SequenceSet` (field `images` is a `Dictionary` of `Dictionary`). Though I admit, I haven't done any profiling to be sure.

Even if these sequence names would be cached in for example `ProductionPaletteWidget` (or even in `BuildableInfo` directly), it would likely still be a `Dictionary` with key containing probably `Player.InternalName` and sequence/image name.

Reply:
I meant that as these collections aren't ordered the system needs to search all entries and there can be loads and loads of them, specifically sequences

Reply:
`Dictionary<TKey, TValue>` is implementation of [Hashtable](https://en.wikipedia.org/wiki/Hashtable). This means that search is O(1) operation, so it's definitely quick (there's no iteration through entire dictionary to find an item). Here's performance comparison between `Array` and `Dictionary`: https://gamedev.center/array-vs-dictionary-lookup-in-net/

Reply:
I suppose it's fine